samples/net/sockets/http_server/src/server.c

```diff
@@ -0,0 +1,889 @@
+				continue;
+			}
+
+			ctx->clients[i - 2].offset += valread;
+			ctx->clients[i - 2].server_state = INITIAL_STATE;
+			handle_http_request(ctx, &ctx->clients[i - 2], i - 2);
+		}
+	} while (ctx->infinite == 1);
+	return 0;
+}
+
+void close_client_connection(
+	struct http2_server_ctx *ctx_server, int client_index)
+{
+	client_index += 2;
+	close(ctx_server->fds[client_index].fd);
+	ctx_server->fds[client_index].fd = 0;
+	ctx_server->fds[client_index].events = 0;
+	ctx_server->fds[client_index].revents = 0;
+
+	if (client_index == ctx_server->num_clients) {
+		while (ctx_server->num_clients > 0 &&
+			ctx_server->fds[ctx_server->num_clients].fd == 0) {
+			--ctx_server->num_clients;
+		}
+	}
+}
+
+int handle_http_request(struct http2_server_ctx *ctx_server,
+	struct http2_client_ctx *ctx_client, int client_index)
+{
+	int ret = -EINVAL;
+
+	do {
+		switch (ctx_client->server_state) {
+		case INITIAL_STATE:
+			ret = handle_initial_state(ctx_client);
+			break;
+		case HTTP_PREFACE:
+			ret = handle_http_preface(ctx_client);
+			break;
+		case HTTP1_REQUEST:
+			ret = handle_http1_request(
+				ctx_server, ctx_client, client_index);
+			break;
+		case HTTP2_FRAME_SETTINGS:
+			ret = handle_http2_frame_setting(ctx_client);
+			break;
+		case HTTP2_FRAME_WINDOW_UPDATE:
+			ret = handle_http2_frame_window_update(ctx_client);
+			break;
+		case STREAMING:
+			ret = handle_http2_streaming(ctx_client);
+			break;
+		case HTTP2_FRAME_GOAWAY:
+			ret = handle_http2_frame_goaway(
+				ctx_server, ctx_client, client_index);
+			break;
+		case CLOSING:
+			break;
+		default:
+			LOG_ERR("Unknown state.\n");
+			close_client_connection(ctx_server, client_index);
+		}
+
+		if (ctx_client->server_state == CLOSING)
+			break;
+
+	} while (1);
+
+	return ret;
+}
+
+int handle_http2_streaming(struct http2_client_ctx *ctx_client)
+{
+	printf("STREAMING\n");
+
+	ctx_client->streams[0].stream_state =
+		determine_stream_state(ctx_client->buffer);
+	int ret = -EINVAL;
+
+	switch (ctx_client->streams[0].stream_state) {
+	case IDLE_STATE:
+		ret = handle_http2_idle_state(ctx_client);
+		break;
+	case OPEN_STATE:
+		ret = handle_http2_open_state(ctx_client);
+		break;
+	case CLOSE_STATE:
+		break;
+	default:
+		LOG_ERR("Unknown state.\n");
+	}
+
+	return 0;
+}
+
+int handle_http_preface(struct http2_client_ctx *ctx_client)
+{
+	if (ctx_client->offset < strlen(preface)) {
+		/* We don't have full preface yet, get more data. */
+		return -EAGAIN;
+	}
+
+	if (strncmp(ctx_client->buffer, preface, strlen(preface)) != 0) {
+		printf("HTTP1_REQUEST.\n");
+		ctx_client->server_state = HTTP1_REQUEST;
+	} else {
+		ctx_client->is_http2 = 1;
+		ctx_client->offset -= strlen(preface);
+		memmove(ctx_client->buffer,
+			ctx_client->buffer + strlen(preface),
+			ctx_client->offset);
+
+		if (ctx_client->offset == 0) {
+			ctx_client->server_state = CLOSING;
+		} else {
+			ctx_client->server_state =
+				determine_server_state(ctx_client->buffer);
```

Comment:
`determine_server_state()` will reach to the 4th byte of the buffer (frame type), but we don't know if there's enough data in the buffer (you didn't check if (offset > 4). That's why in my initial proposal I suggested to have a separate state for parsing FRAME_HEADER, in which we'd evaluate if we have enough data to parse the entire header, and then could decide what to do next.